test/e2e/framework/service/jig.go

```diff
@@ -260,21 +260,34 @@ func (j *TestJig) CreateLoadBalancerService(timeout time.Duration, tweak func(sv
 // GetEndpointNodes returns a map of nodenames:external-ip on which the
 // endpoints of the Service are running.
 func (j *TestJig) GetEndpointNodes() (map[string][]string, error) {
-	nodes, err := e2enode.GetBoundedReadySchedulableNodes(j.Client, MaxNodesForEndpointsTests)
+	nodes, err := j.ListNodesWithEndpoint()
 	if err != nil {
 		return nil, err
 	}
-	epNodes, err := j.GetEndpointNodeNames()
+	nodeMap := map[string][]string{}
+	for _, node := range nodes {
+		nodeMap[node.Name] = e2enode.GetAddresses(node, v1.NodeExternalIP)
+	}
+	return nodeMap, nil
+}
+
+// ListNodesWithEndpoint returns a list of nodes on which the
+// endpoints of the given Service are running.
+func (j *TestJig) ListNodesWithEndpoint() ([]*v1.Node, error) {
+	nodeNames, err := j.GetEndpointNodeNames()
 	if err != nil {
 		return nil, err
 	}
-	nodeMap := map[string][]string{}
-	for _, n := range nodes.Items {
-		if epNodes.Has(n.Name) {
-			nodeMap[n.Name] = e2enode.GetAddresses(&n, v1.NodeExternalIP)
+	ctx := context.TODO()
+	nodes := make([]*v1.Node, 0, nodeNames.Len())
+	for name := range nodeNames {
+		n, err := j.Client.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
```

Comment:
Instead of getting N endpoints individually, you should list all endpoints instead.

Reply:
but all I get here is only the node names and I want to return a list of `*v1.Node`. should I filter the nodes by label `kuberntes.io/hostname`?

Reply:
No - let's list everything.
You may not need everything and that's fine.

Reply:
I mean - list all nodes once and then iterate over them.